src/zope/interface/declarations.py

```diff
@@ -223,9 +223,6 @@ def __cmp(self, other):
         # This spelling works under Python3, which doesn't have cmp().
         return (n1 > n2) - (n1 < n2)
 
-    def __hash__(self):
-        return Declaration.__hash__(self)
-
```

Comment:
Maybe add a few words to the following comment block about also needed hashing to be identity based so that if we need to return to something like this in the future we're not confused by the `TypeError`? Maybe something like (additions in bold)

> We want equality **and hashing** to be based on identity. … (2) the default equality **and hashing** semantics being identity based.